runtime/src/accounts.rs

```diff
@@ -314,36 +346,68 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, (account, None));
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &mut HashMap<Pubkey, Arc<AtomicU64>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
-    ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+    ) -> Result<CreditOnlyLocks> {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            if locks.contains(k) || credit_only_locks.contains_key(k) {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            }
+        }
+        for k in credit_only_keys.iter() {
             if locks.contains(k) {
                 error_counters.account_in_use += 1;
                 debug!("Account in use: {:?}", k);
                 return Err(TransactionError::AccountInUse);
             }
         }
-        for k in keys {
-            locks.insert(**k);
+
+        for k in credit_debit_keys {
+            locks.insert(*k);
+        }
+        let mut lock_cache: CreditOnlyLocks = vec![];
+        for k in credit_only_keys {
+            if !credit_only_locks.contains_key(k) {
+                credit_only_locks.insert(*k, Arc::new(AtomicU64::new(0)));
+            }
+            let lock = credit_only_locks.get(k).unwrap().clone();
+            lock_cache.push(lock);
         }
-        Ok(())
+
+        Ok(lock_cache)
     }
 
-    fn unlock_account(tx: &Transaction, result: &Result<()>, locks: &mut HashSet<Pubkey>) {
+    fn unlock_account(
+        tx: &Transaction,
+        result: &Result<()>,
+        locks: &mut HashSet<Pubkey>,
+        credit_only_locks: &mut HashMap<Pubkey, Arc<AtomicU64>>,
+    ) {
+        let (credit_debit_keys, credit_only_keys) = &tx.message().get_account_keys_by_lock_type();
         match result {
             Err(TransactionError::AccountInUse) => (),
             _ => {
-                for k in &tx.message().account_keys {
+                for k in credit_debit_keys {
                     locks.remove(k);
                 }
+                for k in credit_only_keys {
+                    if let Some(lock) = credit_only_locks.get(k) {
+                        if Arc::strong_count(lock) == 1 {
```

Comment:
Without further synchronization, this check seems racy.  For instance, each of the the last two threads holding two references to an account could both call unlock_account(), both observe that the strong_count() is 2, and so the credit_only_locks is not removed. We might be able to get rid of this ref counting altogether if we decide to only store credit-only accounts at the end . of a slot, when we call freeze()